[doc] Update docs about kernels and functions (#4044)
taichi-dev/taichi: docs/lang/articles/basic/syntax.md

# Kernels and functions

> Taichi has two types of functions: Taichi kernel, and Taichi functions.

Reviewer: ```suggestion
Taichi has two types of functions: Taichi kernels, and Taichi functions.
```

> Scope inside Taichi kernels and Taichi functions is called Taichi scope, and scope outside it is called Python scope.

Reviewer: ```suggestion
Scope inside Taichi kernels and Taichi functions is called Taichi scope, and scope outside them is called Python scope.
```

> A Taichi kernel is the entrypoint of a Taichi program, and it is similar to `__global__` function in CUDA. It can only be called inside Python scope.

Reviewer: ```suggestion
A Taichi kernel is the entrypoint of a Taichi program, and it is similar to a `__global__` function in CUDA. It can only be called inside Python scope.
```

> A Taichi function can only be called inside Taichi scope, and it is similar to `__device__` function in CUDA.

Reviewer: ```suggestion
A Taichi function can only be called inside Taichi scope, and it is similar to a `__device__` function in CUDA.
```

> Major differences of Taichi kernels and Taichi functions are listed in the table below.

Reviewer: ```suggestion
Major differences between Taichi kernels and Taichi functions are listed in the table below.
```

> | | Taichi kernels | Taichi functions |

Reviewer 2: 这儿多一个 | ？

Author: 最左边那列是不同点，就让最上面那行空着了

> | :--- | :---: | :---: |

Reviewer 2: 这里少一个冒号吗？

| Can be called in | Python scope | Taichi scope |
| Argument type annotation | Mandatory | Recommended |
| Return type annotation | Mandatory| Recommended |
| Return value | Scalar/Vector/Matrix | Arbitrary |
| Max number of total elements in arguments | 8 (for OpenGL and CC) or 64 (other) | Unlimited |
| Max number of return values in a return statement | 1 | Unlimited |

> | Max number of elements in return value | 30 | Unlimited |

Reviewer: ```suggestion
| Max number of total elements in return values | 30 | Unlimited |
```

```

:::note

> Taichi supports scalars, `ti.Matrix` and`ti.Vector` as kernel arguments.

Reviewer 3: For `ti.Matrix` and `ti.Vector` , now we only pass it by value. It is different from python `list` and c++ `array`. Maybe we should mention it?

Author: IIRC we support passing `ti.Matrix` and `ti.Vector` by reference by adding type annotation `ti.template()`, and anything passed by reference is inlined into the kernel during transformation. So, maybe we needn't mention it specifically because it is already stated in https://docs.taichi.graphics/lang/articles/advanced/meta#template-metaprogramming ?

Reviewer: Regarding this, I think we need to mention that passing Python objects into kernels with `ti.template()` are read-only.

Author: > Regarding this, I think we need to mention that passing Python objects into kernels with `ti.template()` are read-only because it is basically instantiating templates.

In the current implementation, it seems we allow redefinitions of `ti.template` arguments once because the template arguments are placed in the global variable dict. The previous variable is shadowed by the new variable created which is in the variable dict of transformer. Should we ban this?

Reviewer: Yes we should ban this.

The total number of elements in kernel arguments must not exceed 8 on OpenGL and CC backends, or 64 on other backends.

> The element number of a scalar argument is 1, and the element number of a `ti.Matrix` or`ti.Vector` is the number of elements inside it.

Reviewer: ```suggestion
The number of elements in a scalar argument is 1, and the number of elements in a `ti.Matrix` or`ti.Vector` is the number of elements inside it.
```
The latter part of this sentence appears meaningless. Is there a better way to emphasize this?
